joblib/numpy_pickle.py

```diff
@@ -446,22 +336,36 @@ def load(filename, mmap_mode=None):
     file was saved with compression, the arrays cannot be memmaped.
     """
     with open(filename, 'rb') as file_handle:
-        # We are careful to open the file handle early and keep it open to
-        # avoid race-conditions on renames. That said, if data are stored in
-        # companion files, moving the directory will create a race when
-        # joblib tries to access the companion files.
-        if _read_magic(file_handle) == _ZFILE_PREFIX:
-            if mmap_mode is not None:
-                warnings.warn('file "%(filename)s" appears to be a zip, '
-                              'ignoring mmap_mode "%(mmap_mode)s" flag passed'
-                              % locals(), Warning, stacklevel=2)
-            unpickler = ZipNumpyUnpickler(filename, file_handle=file_handle)
-        else:
-            unpickler = NumpyUnpickler(filename, file_handle=file_handle,
-                                       mmap_mode=mmap_mode)
+        magic = _read_magic(file_handle)
+
+    # Backward compatibility with old compression strategy
+    if magic == _ZFILE_PREFIX:
+        warnings.warn("You are using an old version of joblib cache. Please "
+                      "regenerate your cache.",
+                      DeprecationWarning, stacklevel=2)
+        return load_compatibility(filename)
+
+    with closing(_check_filetype(filename, magic)) as file_handle:
+        if isinstance(file_handle, gzip.GzipFile) and mmap_mode is not None:
+            warnings.warn('File "%(filename)s" appears to be compressed, '
+                          'this is not compatible with mmap_mode '
+                          '"%(mmap_mode)s" flag passed' % locals(),
+                          DeprecationWarning, stacklevel=2)
 
+        # We are careful to open the file handle early and keep it open to
+        # avoid race-conditions on renames.
+        # That said, if data are stored in companion files, which can be the
+        # case with the old persistence format, moving the directory will
+        # create a race when joblib tries to access the companion files.
+        unpickler = NumpyUnpickler(filename,
+                                   file_handle,
+                                   mmap_mode=mmap_mode)
         try:
             obj = unpickler.load()
+            if unpickler.compat_mode:
+                warnings.warn("You are using an old version of joblib cache. "
```

Comment:
This is not really the cache you are talking about but more pickles. In my mind "caching" in a joblib context means joblib.Memory.

You have this slight issue in a few other places, e.g. CHANGES.rst.